stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -0,0 +1,720 @@
+                    "FATAL: unreachable: burnchain block height has no reward cycle"
+                ));
+
+            let last_processed_reward_cycle = {
+                let ic = self.sortition_db.index_handle(&canonical_sortition_tip);
+                ic.get_last_processed_reward_cycle()?
+            };
+
+            if last_processed_reward_cycle > current_reward_cycle {
+                // already processed upcoming reward cycle
+                continue;
+            }
+
+            // This is the first Stacks block in the prepare phase for the next reward cycle.
+            // Pause here and process the next sortitions
+            return Ok(Some(block_hash));
+        }
+
+        // no PoX anchor block found
+        Ok(None)
+    }
+
+    /// Given a burnchain header, find the PoX reward cycle info
+    pub fn get_nakamoto_reward_cycle_info(
+        &mut self,
+        burn_header: &BurnchainBlockHeader,
+    ) -> Result<Option<RewardCycleInfo>, Error> {
+        let sortition_tip_id = self
+            .canonical_sortition_tip
+            .as_ref()
+            .expect("FATAL: Processing anchor block, but no known sortition tip");
+
+        get_nakamoto_reward_cycle_info(
+            burn_header.block_height,
+            sortition_tip_id,
+            &self.burnchain,
+            &mut self.chain_state_db,
+            &self.sortition_db,
+            &self.reward_set_provider,
+        )
+    }
+
+    /// Process the next-available burnchain block, if possible.
+    /// Burnchain blocks can only be processed for the last-known PoX reward set, which is to say,
+    /// burnchain block processing can be blocked on the unavailability of the next PoX anchor
+    /// block.  If the next PoX anchor block is not available, then no burnchain block processing
+    /// happens, and the hash of the PoX anchor block is returned instead.
+    ///
+    /// Returns Ok(None) if all burnchain blocks are processed
+    /// Returns Ok(Some(hash)) if burnchain block processing is blocked on a missing PoX anchor
+    /// block
+    /// Returns Err(..) if an error occurred while processing (i.e. a DB error).
+    pub fn handle_new_nakamoto_burnchain_block(
+        &mut self,
+    ) -> Result<Option<BlockHeaderHash>, Error> {
+        // highest burnchain block we've downloaded
+        let canonical_burnchain_tip = self.burnchain_blocks_db.get_canonical_chain_tip()?;
+
+        debug!("Handle new canonical burnchain tip";
+               "height" => %canonical_burnchain_tip.block_height,
+               "block_hash" => %canonical_burnchain_tip.block_hash.to_string());
+
+        // Retrieve all the direct ancestors of this block with an unprocessed sortition
+        let mut cursor = canonical_burnchain_tip.block_hash.clone();
+        let mut sortitions_to_process = VecDeque::new();
+
+        // We halt the ancestry research as soon as we find a processed parent
+        let mut last_processed_ancestor = loop {
```

Comment:
NIT: This is by no means important, but it might be nice to pull this loop into a function called "get_last_processed_ancestor" as this function is quite long and am having a hard time following it. Am sure there are other logical blocks that might be better in a sep function, but just commenting on this one to spur thought about it.

Reply:
Ha, this is copypasta from the epoch 2.x coordinator.  I'll consolidate.